src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -577,48 +664,53 @@ private boolean isInRange(DecoratedKey key)
             return null;
         }
 
-        public UnfilteredRowIterator readAndValidatePartition(PrimaryKey key, List<PrimaryKeyWithSortKey> primaryKeys)
+        public UnfilteredRowIterator validatePartition(PrimaryKey key, List<PrimaryKeyWithSortKey> primaryKeys, UnfilteredRowIterator partition)
         {
-            // If we've already processed the key, we can skip it. Because the score ordered iterator does not
-            // deduplicate rows, we could see dupes if a row is in the ordering index multiple times. This happens
-            // in the case of dupes and of overwrites.
-            if (processedKeys.contains(key))
-                return null;
+            queryContext.addPartitionsRead(1);
+            queryContext.checkpoint();
+            var staticRow = partition.staticRow();
+            UnfilteredRowIterator clusters = applyIndexFilter(partition, filterTree, queryContext);
 
-            try (UnfilteredRowIterator partition = controller.getPartition(key, view, executionController))
+            if (clusters == null || !clusters.hasNext())
             {
-                queryContext.addPartitionsRead(1);
-                queryContext.checkpoint();
-                var staticRow = partition.staticRow();
-                UnfilteredRowIterator clusters = applyIndexFilter(partition, filterTree, queryContext);
-
-                if (clusters == null || !clusters.hasNext())
-                {
-                    processedKeys.add(key);
-                    return null;
-                }
+                processedKeys.put(key, Boolean.TRUE);
+                return null;
+            }
 
-                var now = FBUtilities.nowInSeconds();
-                boolean isRowValid = false;
-                var row = clusters.next();
-                assert !clusters.hasNext() : "Expected only one row per partition";
-                if (!row.isRangeTombstoneMarker())
+            var now = FBUtilities.nowInSeconds();
+            boolean isRowValid = false;
+            var row = clusters.next();
+            assert !clusters.hasNext() : "Expected only one row per partition";
```

Comment:
Perhaps this method could be named `validateSingleRowPartition` for more clarity?